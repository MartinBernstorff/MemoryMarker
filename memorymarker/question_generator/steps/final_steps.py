from dataclasses import dataclass
from typing import TYPE_CHECKING

from instructor import patch
from iterpy.iter import Iter  # noqa: TCH002
from openai import AsyncOpenAI
from openai.types.chat.chat_completion_user_message_param import (
    ChatCompletionUserMessageParam,
)

from memorymarker.question_generator.qa_prompt import QAPromptResponseModel, QAResponses

if TYPE_CHECKING:
    from openai.types.chat.chat_completion_message_param import (
        ChatCompletionMessageParam,
    )


@dataclass
class QuestionExtractor:
    openai_api_key: str
    name: str = "question-extractor"
    model = "gpt-4-0125-preview"

    def __post_init__(self):
        self.client = patch(  # type: ignore # Incorrectly typed from Instructor library
            AsyncOpenAI(api_key=self.openai_api_key)
        )

    def _build_message(self, input_str: str) -> "ChatCompletionMessageParam":
        return ChatCompletionUserMessageParam(
            role="user",
            name="UserName",
            content=f"""This is a set of questions and answers. Extract them to the following model. {input_str}""",
        )

    async def __call__(self, input_str: str) -> Iter["QAPromptResponseModel"]:
        return await self.client.chat.completions.create(
            model=self.model,
            messages=[self._build_message(input_str=input_str)],
            response_model=QAResponses,
            temperature=0.0,
        )  # type: ignore
