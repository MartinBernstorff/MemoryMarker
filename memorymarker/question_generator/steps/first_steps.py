from dataclasses import dataclass
from typing import TYPE_CHECKING

from memorymarker.question_generator.expanded_pipeline import Model, PipelineFirstStep

if TYPE_CHECKING:
    from memorymarker.document_providers.contextualized_highlight import (
        ContextualizedHighlight,
    )


@dataclass
class COT(PipelineFirstStep):
    model: Model
    variant: str

    async def __call__(self, highlight: "ContextualizedHighlight") -> str:
        base_prompt = f"""The following is a highlight from a document called "{highlight.source_doc_title}".

Why might a student highlight this sentence? Focus on them wanting to learn something, especially if they can apply it in a situation. Think step by step.

Highlight: {highlight.highlighted_text}
"""
        return base_prompt
