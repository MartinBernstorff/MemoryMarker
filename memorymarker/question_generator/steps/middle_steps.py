from dataclasses import dataclass

from memorymarker.question_generator.expanded_pipeline import Model, PipelineMiddleStep


@dataclass(frozen=True)
class StudentReflection(PipelineMiddleStep):
    model: Model
    base_name: str = "student-reflection"

    async def __call__(self, input_text: str) -> str:
        prompt = f"""A student is interested in learning the followign reasoning:

{input_text}

Write one or more questions intended to get a student to reflect on these points. Questions should include:

* Comparison to alternatives
* Pros and cons about the approach
* Example cases for reflection

For each question, also write a one-sentence answer."""

        return await self.model.create_completion(message=prompt)
