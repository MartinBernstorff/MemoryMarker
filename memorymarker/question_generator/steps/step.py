from typing import TYPE_CHECKING, Protocol

if TYPE_CHECKING:
    import pydantic

    from memorymarker.question_generator.completers.openai_completer import (
        ModelCompleter,
    )
    from memorymarker.question_generator.reasoned_highlight import Highlights


class FlowStep(Protocol):
    def identity(self) -> str:
        ...

    async def __call__(self, highlight: "Highlights") -> "Highlights":
        ...


class ResponseModelStep:
    model_completer: "ModelCompleter"

    async def __call__(self, highlight: "Highlights") -> "pydantic.BaseModel":
        ...
