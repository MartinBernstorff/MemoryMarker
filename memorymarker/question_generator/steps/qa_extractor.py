from dataclasses import dataclass
from typing import TYPE_CHECKING

from memorymarker.question_generator.steps.step import FlowStep

if TYPE_CHECKING:
    from memorymarker.question_generator.completers.openai_completer import (
        ModelCompleter,
    )
    from memorymarker.question_generator.qa_responses import QAResponses
    from memorymarker.question_generator.reasoned_highlight import Highlights


@dataclass(frozen=True)
class QuestionExtractionStep(FlowStep):
    completer: "ModelCompleter"

    def identity(self) -> str:
        return f"{self.__class__.__name__}_{self.completer.identity()}"

    async def __call__(self, reasoned_highlight: "Highlights") -> "Highlights":
        responses: QAResponses = await self.completer(
            f"""Extract:
{reasoned_highlight.qa_string}
"""
        )  # type: ignore

        reasoned_highlight.question_answer_pairs = [
            item.to_qaprompt(reasoned_highlight) for item in responses.items
        ]
        return reasoned_highlight
