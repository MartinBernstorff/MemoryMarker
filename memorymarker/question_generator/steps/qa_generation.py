from dataclasses import dataclass
from typing import TYPE_CHECKING

from memorymarker.question_generator.steps.step import FlowStep

if TYPE_CHECKING:
    from memorymarker.question_generator.completers.completer import Completer
    from memorymarker.question_generator.reasoned_highlight import ReasonedHighlight


@dataclass(frozen=True)
class QuestionGenerationStep(FlowStep):
    completer: "Completer"
    n_questions: tuple[int, int] = (1, 5)
    prompt = """You are generating interesting questions. The questions should:
* Create understanding: Prompts should foster understanding, activating related concepts and comparing them to one another. Focus especially on why concepts are related to each other, and how they differ. Feel free to ask questions that are challenging or that require deep thought.
* Be concise: Be as short as possible. Be clear, direct and brief.
* Be paraphrased: The questions should never quote the context.
* Be context-independent: The question must cue or supply all context that is necessary to understand the question. Questions must not assume that th user has read the text that generated the question. It should not address the text or use the context of the text in any way.

Document: "{document_title}"

{reasoning}

Please generate between {lower_bound} and {upper_bound} question/answer pairs to reinforce understanding of the concept.

They should be formatted like:

Q. What is the meaning of life?
A. 42
"""

    def identity(self) -> str:
        return f"{self.__class__.__name__}_{self.completer.identity()}"

    async def __call__(self, highlight: "ReasonedHighlight") -> "ReasonedHighlight":
        prompt = self.prompt.format(
            document_title=highlight.source_document.title,
            reasoning=f"""{highlight.reasoning}""" if highlight.reasoning else "",
            lower_bound=self.n_questions[0],
            upper_bound=self.n_questions[1],
        )

        response = await self.completer(prompt)
        highlight.qa_string = response
        highlight.reasoning_prompt = prompt

        return highlight
