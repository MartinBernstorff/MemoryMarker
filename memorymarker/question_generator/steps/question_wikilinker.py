import asyncio
from dataclasses import dataclass
from typing import TYPE_CHECKING

from memorymarker.question_generator.steps.step import FlowStep

if TYPE_CHECKING:
    from memorymarker.question_generator.completers.completer import Completer
    from memorymarker.question_generator.qa_responses import QAPrompt
    from memorymarker.question_generator.reasoned_highlight import Highlights


@dataclass(frozen=True)
class QuestionWikilinker(FlowStep):
    completer: "Completer"
    prompt = """In the following, identify the important, domain-specific terms. Then, capitalise them, and surround them with wikilinks. There can be more than one important term. Identify terms as you would in a wikipedia article.

E.g.:
When working with version control, why is the git amend command misleading?

Turns into:
When working with [[Version control]], why is the [[Git amend]] command misleading?
Here is the question:
{question}
"""

    def identity(self) -> str:
        return f"{self.__class__.__name__}_{self.completer.identity()}"

    async def _wikilink_prompt(self, question: "QAPrompt") -> "QAPrompt":
        prompt = self.prompt.format(question=question.question)
        response = await self.completer(prompt)  # type: ignore
        question.question = response
        return question

    async def __call__(self, highlight: "Highlights") -> "Highlights":
        prompts = highlight.question_answer_pairs
        wikilinked_prompts = await asyncio.gather(
            *[self._wikilink_prompt(prompt) for prompt in prompts]
        )
        highlight.question_answer_pairs = wikilinked_prompts
        return highlight
