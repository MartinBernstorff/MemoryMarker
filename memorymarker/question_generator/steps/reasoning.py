from dataclasses import dataclass
from typing import TYPE_CHECKING

from memorymarker.question_generator.steps.step import FlowStep

if TYPE_CHECKING:
    from memorymarker.question_generator.completers.openai_completer import Completer
    from memorymarker.question_generator.reasoned_highlight import ReasonedHighlight


@dataclass(frozen=True)
class ReasoningStep(FlowStep):
    completer: "Completer"
    prompt = """You are a teacher at a university, helping students understand what and why a concept is important.

This is a highlight from a document titled {document_title}, which the student found important.

{context}

Think through why the student should be interested in this concept, and what they can learn from it. Think step by step, one bullet point at a time, with at least 5 bullet points. Each bullet point should be brief.

Ask "Why is that?" at the beginning of each bullet point.
"""

    def identity(self) -> str:
        return f"{self.__class__.__name__}_{self.completer.identity()}"

    async def __call__(self, highlight: "ReasonedHighlight") -> "ReasonedHighlight":
        prompt = self.prompt.format(
            document_title=highlight.source_document.title,
            context=highlight.highlighted_text,
        )

        reasoning = await self.completer(prompt)
        highlight.reasoning = reasoning
        highlight.reasoning_prompt = prompt

        return highlight
