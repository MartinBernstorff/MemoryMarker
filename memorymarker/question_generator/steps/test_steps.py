import os
from dataclasses import dataclass, field
from typing import TYPE_CHECKING, Callable, Type

import pytest

from memorymarker.question_generator.completers.openai_completer import (
    OpenAICompleter,
    OpenAIModelCompleter,
)
from memorymarker.question_generator.qa_responses import QAResponses
from memorymarker.question_generator.steps.qa_extractor import QuestionExtractionStep
from memorymarker.question_generator.steps.qa_generation import QuestionGenerationStep
from memorymarker.question_generator.steps.reasoning import ReasoningStep

if TYPE_CHECKING:
    from memorymarker.question_generator.steps.step import FlowStep


def gpt_4_completer() -> OpenAICompleter:
    return OpenAICompleter(
        api_key=os.getenv("OPENAI_API_KEY", "No OPENAI_API_KEY in environment"),
        model="gpt-4-turbo-preview",
    )


@dataclass
class Step:
    step: Type["FlowStep"]
    completer: OpenAICompleter | OpenAIModelCompleter = field(
        default_factory=gpt_4_completer
    )

    def __call__(self) -> "FlowStep":
        return self.step(completer=self.completer)


@pytest.mark.parametrize(
    ("step"),
    [
        Step(step=ReasoningStep),
        Step(step=QuestionGenerationStep),
        Step(
            step=QuestionExtractionStep,
            completer=OpenAIModelCompleter(
                api_key=os.getenv("OPENAI_API_KEY", ""),
                model="gpt-4-turbo-preview",
                response_model=QAResponses,  # type: ignore
            ),
        ),
    ],
)
def test_step_hashes_are_constant(step: Step, snapshot: str):
    assert step().identity() == snapshot
