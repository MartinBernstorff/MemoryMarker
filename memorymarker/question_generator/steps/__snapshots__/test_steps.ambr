# serializer version: 1
# name: test_step_hashes_are_constant[step0]
  'ReasoningStep_OpenAICompleter_gpt-4-turbo-preview'
# ---
# name: test_step_hashes_are_constant[step1]
  'QuestionGenerationStep_OpenAICompleter_gpt-4-turbo-preview'
# ---
# name: test_step_hashes_are_constant[step2]
  'QuestionExtractionStep_OpenAIModelCompleter_gpt-4-turbo-preview'
# ---
